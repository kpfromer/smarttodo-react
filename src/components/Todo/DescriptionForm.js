import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class DescriptionForm extends Component {

  static propTypes = {
    onUpdate: PropTypes.func.isRequired,
    description: PropTypes.string,
    className: PropTypes.string,
    invalidClassName: PropTypes.string,
    clearOnUpdate: PropTypes.bool
  };

  static defaultProps = {
    description: '',
    className: '',
    invalidClassName: '',
    clearOnUpdate: false
  };

  state = {
    description: this.props.description,
    invalid: false
  };

  updateDescription = event => {
    const value = event.target.value;
    this.setState({
      description: value,
      invalid: value.length === 0
    });
  };

  handleSubmit = event => {
    event.preventDefault();
    const { description } = this.state;
    if (description.length > 0) {
      this.props.onUpdate(description);
      if (this.props.clearOnUpdate) {
        this.setState({ description: '' });
      }
    } else {
      this.setState({ invalid: true })
    }
  };

  render() {
    const { description, invalid } = this.state;
    const { className, invalidClassName } = this.props;
    const invalidStyle = invalid ? invalidClassName : '';
    return (
      <form onSubmit={this.handleSubmit}>
        <input
          className={`${className} ${invalidStyle}`}
          value={description}
          onChange={this.updateDescription}
        />
      </form>
    );
  }
}
