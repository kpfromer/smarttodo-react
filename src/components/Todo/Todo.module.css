.todo {
  position: relative;
  font-size: 24px;
  border-bottom: 1px solid #ededed;
}

.todo:last-child {
  border-bottom: none;
}

.todo.editing {
  border-bottom: none;
  padding: 0;
}

.todo.editing .edit {
  display: block;
  width: 506px;
  padding: 12px 16px;
  margin: 0 0 0 43px;
}

.todo.editing .view {
  display: none;
}

.todo .toggle {
  text-align: center;
  width: 40px;
  /* auto, since non-WebKit browsers doesn't support input styling */
  height: auto;
  position: absolute;
  top: 0;
  bottom: 0;
  margin: auto 0;
  border: none; /* Mobile Safari */
}

.todo .toggle {
  opacity: 0;
}

.todo .toggle + label {
  /*
      Firefox requires `#` to be escaped - https://bugzilla.mozilla.org/show_bug.cgi?id=922433
      IE and Edge requires *everything* to be escaped to render, so we do that instead of just the `#` - https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/7157459/
  */
  background: url('data:image/svg+xml;utf8,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2240%22%20height%3D%2240%22%20viewBox%3D%22-10%20-18%20100%20135%22%3E%3Ccircle%20cx%3D%2250%22%20cy%3D%2250%22%20r%3D%2250%22%20fill%3D%22none%22%20stroke%3D%22%23ededed%22%20stroke-width%3D%223%22/%3E%3C/svg%3E')
  no-repeat
  center left;
}

.todo .toggle:checked + label {
  background-image: url('data:image/svg+xml;utf8,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2240%22%20height%3D%2240%22%20viewBox%3D%22-10%20-18%20100%20135%22%3E%3Ccircle%20cx%3D%2250%22%20cy%3D%2250%22%20r%3D%2250%22%20fill%3D%22none%22%20stroke%3D%22%23bddad5%22%20stroke-width%3D%223%22/%3E%3Cpath%20fill%3D%22%235dc2af%22%20d%3D%22M72%2025L42%2071%2027%2056l-4%204%2020%2020%2034-52z%22/%3E%3C/svg%3E');
}

.todo label {
  word-break: break-all;
  padding: 15px 15px 15px 60px;
  display: block;
  line-height: 1.2;
  transition: color 0.4s;
}

.todo.completed label {
  color: #d9d9d9;
  text-decoration: line-through;
}

/* TODO: fix padding margin top */
.todo .destroy {
  display: none;
  position: absolute;
  top: 0;
  right: 10px;
  bottom: 0;
  width: 40px;
  height: 40px;
  font-size: 30px;
  color: #cc9a9a;
  margin: auto 0 11px;
  transition: color 0.2s ease-out;
}

.todo .destroy:hover {
  color: #af5b5e;
}

.todo .destroy:after {
  content: 'Ã—';
}

.todo:hover .destroy {
  display: block;
}

.todo .edit {
  display: none;
  transition: border-bottom 250ms;
}

.edit.invalid {
  border-bottom: 1px solid #ff3228;
}

.todo.editing:last-child {
  margin-bottom: -1px;
}