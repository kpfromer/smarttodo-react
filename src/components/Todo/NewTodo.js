import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import DescriptionForm from "./DescriptionForm";

export default class NewTodo extends PureComponent {

  static propTypes = {
    onCreate: PropTypes.func.isRequired
  };

  render() {
    return (
      <ListItem>
        <ListItemText>
          <DescriptionForm clearOnUpdate onUpdate={this.props.onCreate}/>
        </ListItemText>
      </ListItem>
    );
  }
}
