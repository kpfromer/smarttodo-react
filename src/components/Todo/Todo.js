import React, { Component } from "react";
import PropTypes from 'prop-types';

import styles from './Todo.module.scss'

import DescriptionForm from "./DescriptionForm";

export default class Todo extends Component {

  static propTypes = {
    description: PropTypes.string.isRequired,
    completed: PropTypes.bool.isRequired,
    onCheck: PropTypes.func.isRequired,
    onRemove: PropTypes.func.isRequired,
    onUpdate: PropTypes.func.isRequired
  };

  state = {
    editing: false
  };

  checkTodo = event => this.props.onCheck(event.target.checked);
  removeTodo = () => this.props.onRemove();

  startEditing = () =>
    this.setState({ editing: true });

  handleUpdate = description => {
    this.setState({ editing: false });
    this.props.onUpdate(description);
  };

  render() {
    const { description, completed } = this.props;

    const completedStyle = completed ? styles.completed : '';
    const editingStyle = this.state.editing ? styles.editing : '';

    return (
      <li className={`${styles.todo} ${completedStyle} ${editingStyle}`}>
        <div className={styles.view}>
          <input
            className={styles.toggle}
            type="checkbox"
            checked={completed}
            onChange={this.checkTodo}
          />
          <label onDoubleClick={this.startEditing}>{description}</label>
          <button className={styles.destroy} onClick={this.removeTodo}/>
        </div>
        <DescriptionForm
          className={styles.edit}
          invalidClassName={styles.invalid}
          description={description}
          onUpdate={this.handleUpdate}
        />
      </li>
    );
  }
}