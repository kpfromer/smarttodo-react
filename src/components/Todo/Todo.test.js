import { SetupComponent } from "react-component-setup";
import Todo from "./Todo";
import DescriptionForm from "./DescriptionForm";

const { shallow: setup } = SetupComponent({
  Component: Todo,
  defaultProps: {
    id: '666',
    description: 'what todo?',
    completed: true,
    onCheck: () => {},
    onRemove: () => {},
    onUpdate: () => {}
  }
});

describe('Todo', () => {

  it('renders without crashing', () => {
    setup();
  });

  it('renders a checkbox and displays `description` label', () => {
    const { wrapper } = setup();
    expect(wrapper.find('.view')).toMatchSnapshot();
  });

  it('renders the DescriptionForm for editing', () => {
    const { wrapper } = setup();
    expect(wrapper.find(DescriptionForm)).toMatchSnapshot();
  });

  describe('on checkbox click', () => {
    let wrapper, mockCheck;

    beforeEach(() => {
      mockCheck = jest.fn();

      ({ wrapper } = setup({ onCheck: mockCheck }));

      wrapper.find('input').simulate('change', { target: { checked: false } });
    });

    it('runs `onCheck` with id and checked value', () => {
      expect(mockCheck).toHaveBeenCalledWith('666', false);
    });
  });

  describe('on label double click', () => {
    let wrapper, mockUpdate;

    beforeEach(() => {
      mockUpdate = jest.fn();

      ({ wrapper } = setup({
        onUpdate: mockUpdate
      }));

      wrapper.find('label').simulate('doubleClick');
    });

    it('renders Todo with editing `className`', () => {
      expect(wrapper).toHaveClassName('editing');
    });

    describe('on DescriptionForm update', () => {
      beforeEach(() => {
        wrapper.find(DescriptionForm).props().onUpdate('new description');
      });

      it('runs `onUpdate` with id and updated description', () => {
        expect(mockUpdate).toHaveBeenCalledWith('666', 'new description');
      });
    });
  });

  describe('on remove button click', () => {
    let wrapper, mockRemove;

    beforeEach(() => {

      mockRemove = jest.fn();

      ({ wrapper } = setup({
        onRemove: mockRemove
      }));

      wrapper.find('button').simulate('click');
    });

    it('runs `onRemove` with id', () => {
      expect(mockRemove).toHaveBeenCalledWith('666');
    });
  });
});