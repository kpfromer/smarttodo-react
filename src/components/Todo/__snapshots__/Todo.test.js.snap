// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo on ListItemText click renders the DescriptionForm for editing 1`] = `
<DescriptionForm
  clearOnUpdate={false}
  description="what todo?"
  onUpdate={[Function]}
  textFieldProps={
    Object {
      "autoFocus": true,
    }
  }
/>
`;

exports[`Todo renders a checkbox and displays \`description\` label 1`] = `
<WithStyles(ListItem)
  divider={true}
>
  <React.Fragment>
    <WithStyles(Checkbox)
      checked={true}
      onChange={[Function]}
      tabIndex={1}
    />
    <WithStyles(ListItemText)
      onClick={[Function]}
    >
      <FindLinks
        handleLink={[Function]}
      >
        what todo?
      </FindLinks>
    </WithStyles(ListItemText)>
    <WithStyles(ListItemSecondaryAction)>
      <WithStyles(IconButton)
        aria-label="Delete"
        onClick={[Function]}
      >
        <pure(Delete) />
      </WithStyles(IconButton)>
    </WithStyles(ListItemSecondaryAction)>
  </React.Fragment>
</WithStyles(ListItem)>
`;
